Support for reading off-chain data (hotel description and room types)

JirkaChadima opened the pull request:
After https://github.com/windingtree/wt-js-libs/pull/127 and https://github.com/windingtree/wt-js-libs/pull/125 are done, we have to reflect the changes in the API as well. We are starting with the Read part as it looks more clear. The goal of this PR is to implement a `GET /hotel/:address` endpoint that would provide off-chain stored data upon request. We support only a `description` file for now.

**TODO**

- [x] Incorporate changes from https://github.com/windingtree/wt-js-libs/pull/138 and wt-contracts 0.2.2
- [x] Remove data modifying endpoints and tests (will be moved to a separate API)
- [x] Remove wallet managerment (not necessary when we're not modifying the data)
- [x] Reconfigure the whole thing so it actually uses http and swarm adapters and not the in-memory one
- [x] Reponses may contain undocumented `error` field - that's just weird - either document it or drop it
- [x] We should not download documents unless necessary - compute fields and only then download off-chain documents, it will bring a major speedup to the API
- [x] We should enforce checksum address format - if I get a hotel with different letter cases, API blows up with 404, it should respond with a redirect with properly cased address in the checksum format
- [x] Add pagination service
- [x] Naming and code organization can be improved

MatiasOS commented on docs/swagger.json:
```diff
@@ -236,7 +260,7 @@
         ],
         "summary": "Remove Hotel.",
-        "description": "Remove Hotel by contract address.",
+        "description": "Remove Hotel byfrom WT Index.",
```
There is a small typo `byfrom`

MatiasOS commented on docs/swagger.json:
```diff
@@ -71,8 +71,8 @@
         ],
-        "summary": "Create new hotel",
-        "description": "Create a new hotel. Data is wrapped into a hotel property to enable easier extension in the future.",
+        "summary": "Register new hotel in WT Index. This is subject to change.",
```
Using register, it makes me think that I should create the hotel on my own and then register it. 
I know it's quite subjective, but maybe someone else is thinking the same. What do you think?

JirkaChadima replied:
That's actually intentional. I'd say the terminology might be:

- create - lay down all hotel data (description, rooms, etc) in an off-chain storage
- register - write the hotel to WT Index

WDYT @czervenka ?

JirkaChadima commented:
I think you're on the right track @MatiasOS , there's just one thing not clear to me with this solution. How will this work with the asynchronously resolved off-chain data fields?

If you look in the swagger docs, there's the `dataUri` hotel property, that contains a link to an off-chain stored document that holds two properties: `description` and `descriptionUri`. The idea behind includeFields and excludeFields is, that you can actually choose which off chain data you would like to get resolved and thus save time required for their resolving if you don't need them. The previously existing fields such as name and location have moved to an off-chain stored description. So the tests are actually testing an obsolete scenario (which is probably a result of not using the latest (although unreleased) wt-js-libs in `proposal/next`).

So if you would download everything at first and then drop some fields (which seems like the thing you are doing now), it wouldn't really save time. Am I missing something in the code?

JirkaChadima commented:
This looks better, I'll have a proper look at it this afternoon.

JirkaChadima commented:
There's a big big difference between the actual data shape that the API returns and the format designed in the original swagger docs. That unfortunately leads to a kind of weird (at least probably inefficient) implementation of include/excludeFields. I'd fix the shape of data in tests and then it should be clearly visible, what are the culprits of current excludeFields implementation.

MatiasOS commented:
According to the previous week, in terms of separating the API, I have disabled manager tests to migrate the functionality to another repository soon.
Also, I have disable on the transaction tests since they stopped working with the update, to treat it in a later pull request.

please @czervenka , check that the expected result is up to date with what is specified in the documentation.

JirkaChadima commented:
Hey @MatiasOS, We were talking today with @czervenka  about his changes to the API design on Swaggerhub and found a few minor changes and updates. He'll fix those and update the swaggerfile directly in this repo. I'll do a review of the code after that.

As for the manager part (POST, PUT), I'd delete it altogether from here and update README accordingly. Our goal should be to have the swagger docs and code in line. Since there are no manager endpoints in the updated swaggerfile, there is no place for them in the code. We'll revive them later :-)

Also @czervenka, we should delete from the imported swaggerfile in the repo all operations that are currently not implemented (such as `/ratePlans` and `/availability`).

Thanks!

czervenka commented:
I don't know the repository, @JirkaChadima will do the review. I am still finishing [API spec on swaggerhub](https://app.swaggerhub.com/apis/windingtree/api-for-reading). I will add latest version here when done.

JirkaChadima commented:
@MatiasOS OK, so I have a few more notes before I'll get to fix the code organization a little bit:

- eslint still returns some errors
- Update InMemoryAdapter to 3.0.0 (my bad, I forgot to publish that) and update the instantiation of all adapters so that it actually works (the default export got changed and we don't need `.adapter` anymore)
- We're still passing a deprecated `dataModelType` option to wt-js-libs. Drop that.
- the `swarmProviderUrl` option is not actually taken from global config in wt-js-libs config
- checksumAddress - I'd prefer throwing errors instead of silent conversion. It looks like magic is happening for the outside user and there's no motivation for repairing their reference
- off-chain data is still unnecessarily downloaded in `filter-responses`. You don't need `indexRow` and `description` variables at all if there are no descriptionFields. That might be a huge performance degradation.
- The whole `/roomTypes` endpoint does not exist even though it's specified in the documentation
- pagination - If I pass a limit of 1, it doesn't work
- pagination - What's the deal with silent rewriting of attributes? Again, it behaves slightly like magic, I'd very much prefer crashing the whole request. The clients should behave properly, it's not a responsibility of the API to fix their mistakes.

JirkaChadima commented:
@MatiasOS I think I'm happy with the current state. Have a look at my changes and I think we can merge it. I've just simplified and renamed a few things here and there.

MatiasOS commented:
@JirkaChadima I'm ok with the changes. We can merge now! :+1:
